enhancements/network/ingress-node/ingress-node-firewall-with-ebpf.md

```diff
@@ -0,0 +1,284 @@
+For the best performance XDP programs need to attach to interfaces using the following models:
+- Native XDP \
+  The XDP program is loaded by the network card driver as part of its initial receive path. This also requires support from the network card driver.
+- Offloaded XDP \
+  The XDP program loads directly on the NIC, and executes without using the CPU. This also requires support from the network interface device.
+
+## Motivation
+
+Customers would like to secure OCP cluster from external attacks.
+
+### Goals
+
+- Flexible match filter rules for customers to use with a CRD validation to detect and reject invalid configuration.
+- Provide a mechanism for customers to select which nodes via `nodeSelector`.
+- Provide a mechanism to define attachment interfaces for the node ingress firewall objects.  
+- Ingress node firewall is deployed on a fully operational OCP cluster (i.e. Ingress node firewall is deployed as a day 2 operation and is not enabled by default at the cluster install time).
+- fail safe in case of user provisioned invalid rules which can lockup cluster operation.
+
+### Non-Goals
+- Stateful support
+- Flood Attacks
+
+## Proposal
+
+- customers will label nodes where ingress node policy will be deployed on.
+- customers will deploy daemonSet controller on all the selected nodes by applying `IngressNodeFirewallConfig` object. 
+- customers will prepare `IngressNodeFirewall` objects with all required firewall rules and interfaces the ebpf program will be attached to.
+- IngressNodeFW controller will process `IngressNodeFirewall` objects, verification webhook will make sure configuration doesn't contradict with fail safe rules.
+- Ingress node firewall daemon will watch ingress node firewall objects and create `nodeState` object which contains all rules objects that is applied to specific node then load eBPF map table and attach eBPF kernel hook loaded as `BPF_PROG_TYPE_XDP` to the list of interfaces configured in the object.
+- in order to fail safe and avoid cluster lockup, various pinholes will be open in the firewall to allow basic services.
+
+### Workflow Description
+
+- the following diagram shows how node ingress firewall configuration is handled
+
+![Ingress Firewall configuration processing](./ingress-nodefw-config-flow.png)
+
+- the following diagram shows eBPF how ingress firewall kernel hook will process packets and apply firewall rules
+
+![eBPF workflow](./ebpf-packet-flow.png)
+
+- the following table shows sample of fail safe rules to make sure users won't block basic cluster services by configuring invalid firewall rules
+
+| Source CIDR | Protocol | Port | Purpose |
+|-------------|----------|------|---------|
+|  0.0.0.0/0  |   TCP    |  22  |  SSH    |
+|  0.0.0.0/0  |   UDP    |  68  | DHCP    |
+|  0.0.0.0/0  |   TCP    | 6443 | Kubernetes API server access |
+|  0.0.0.0/0  |   TCP    | 2380 & 2379 | Etcd access |
```

Comment:
What exactly does this block? Literally exactly those rules? Can I make one rule that blocks SSH access from 0.0.0.0/1 and one that blocks it from 128.0.0.0/1? Can I make a rule that blocks ports 6442-6444 from 0.0.0.0/0?

Reply:
just show sample of the fail safe rules , and we have webhook in place to prevent overwrite those rules

Reply:
I mean, what do these rules actually mean? What are we actually blocking? It's impossible to know if these are reasonable default without knowing exactly what these rules mean.

Reply:
those rules are set of TCP and UDP ports that customer can't set firewall rules against to keep cluster functionality and we block the user from doing that for example if user attempts to deny TCP using api server port
https://redhat.enterprise.slack.com/files/U01QY8P5YHX/F040RG16DJP/image.png

Reply:
Yes, I know they're a set of rules, but what **_EXACTLY_** do they mean?

For example, if I write a rule saying "block traffic from 0.0.0.0/0 to all ports", then does that (a) get rejected by the admission controller because it would violate all four of the rules written above? (b) get accepted and result in all traffic being blocked except for 0.0.0.0/0 to TCP 22, 0.0.0.0/0 to UDP 68, 0.0.0.0/0 to TCP 6443, 0.0.0.0/0 to TCP 2379, and 0.0.0.0/0 to TCP 2380? (c) get accepted and result in all traffic being blocked to all ports, because none of those 4 rules appears verbatim in the firewall rule?

Reply:
(link doesn't work for me btw)

Reply:
yeah (a) it get rejected because it attempt to overwrite those rules.